src/Transports/MassTransit.Azure.ServiceBus.Core/AzureServiceBusTransport/SessionReceiver.cs

```diff
@@ -9,46 +9,47 @@ namespace MassTransit.AzureServiceBusTransport
     public class SessionReceiver :
         Receiver
     {
-        readonly ClientContext _context;
-        readonly IServiceBusMessageReceiver _messageReceiver;
+        readonly ClientContext _clientContext;
+        readonly ServiceBusReceiveEndpointContext _context;
 
-        public SessionReceiver(ClientContext context, IServiceBusMessageReceiver messageReceiver)
-            : base(context, messageReceiver)
+        public SessionReceiver(ClientContext clientContext, ServiceBusReceiveEndpointContext context)
+            : base(clientContext, context)
         {
+            _clientContext = clientContext;
             _context = context;
-            _messageReceiver = messageReceiver;
         }
 
-        public override Task Start()
+        public override void Start()
         {
-            _context.OnSessionAsync(OnSession, ExceptionHandler);
+            _clientContext.OnSessionAsync(OnSession, ExceptionHandler);
 
-            SetReady();
-
-            return _context.StartAsync();
+            SetReady(_clientContext.StartAsync());
         }
 
         async Task OnSession(ProcessSessionMessageEventArgs messageSession, ServiceBusReceivedMessage message, CancellationToken cancellationToken)
         {
+            var context = new ServiceBusReceiveContext(message, _context);
+            MessageLockContext lockContext = new ServiceBusSessionMessageLockContext(messageSession, message);
+
+            CancellationTokenRegistration registration = default;
+            if (cancellationToken.CanBeCanceled)
+                registration = cancellationToken.Register(context.Cancel);
+
             try
             {
-                await _messageReceiver.Handle(message, cancellationToken, context => AddReceiveContextPayloads(context, messageSession, message))
-                    .ConfigureAwait(false);
+                context.GetOrAddPayload<MessageSessionContext>(() => new ServiceBusMessageSessionContext(messageSession));
```

Comment:
same changes from base class here, I'd rather duplicate the logic to make it efficient vs having this spread out all over the place.